### ISSUE TITLE
question about "connect"

### ISSUE THREAD

**Issue by kolonse** [noodlefrenzy/node-amqp10#190]
Hi,I found a question.If I close the amqp server,such as ActiveMQ, and start my client,then start ActiveMQ server, I found this code to run:
```  
self._session.once(Session.Mapped, function(s) {
  debug('mapped');
  resolve(self);
});
```
but in my code not run to then method. i guess it is promise problem, so I write a test case for promise,like this:
```
var Promise = require("promise");
function t(){
  console.log("test");
}

function test(){
  return new Promise(function(fulfill,reject){
    setInterval(function(){
      fulfill(true);
    },1000);
  });
}

test().then(t)
```

the result only output "test" once.
Do you understand me?

**Comment by mbroadst** [comment]
@kolonse are you saying that when you run this code, you never get to the `console.log('do something here');`:

```
var AmqpClient = require('amqp10').Client;
var client = new AmqpClient();
client.connect('amqp://<your server>')
  .then(function() { 
    console.log('do something here');     // <--- does this ever happen?
  });
```

If that is the case, could you try running that script with the `DEBUG=amqp*` environment variable set?

**Comment by kolonse** [comment]
@mbroadst   Yes! I tried it!  Never get to console.log('do something here');  But get to debug('mapped');

**Comment by mbroadst** [comment]
@kolonse would it be possible for you to provide a trace log using the environment variable "DEBUG=amqp*" before your script? This will help us better diagnose what's happening

**Comment by kolonse** [comment]
@mbroadst  here is my print:
"C:\Program Files (x86)\JetBrains\WebStorm 8.0.3\bin\runnerw.exe" C:\node\node.exe testLogClient.js
Fri, 20 Nov 2015 02:49:35 GMT amqp10:client connecting to: amqp://localhost
run
Fri, 20 Nov 2015 02:49:35 GMT amqp10:connection Connecting to localhost:5672 via straight-up sockets
Fri, 20 Nov 2015 02:49:35 GMT amqp10:connection Transitioning from DISCONNECTED to START due to connect
Fri, 20 Nov 2015 02:49:36 GMT amqp10:connection Error from socket: Error: connect ECONNREFUSED
Fri, 20 Nov 2015 02:49:36 GMT amqp10:client connection error:  { [Error: connect ECONNREFUSED]
  code: 'ECONNREFUSED',
  errno: 'ECONNREFUSED',
  syscall: 'connect' }
Fri, 20 Nov 2015 02:49:36 GMT amqp10:client disconnected
Fri, 20 Nov 2015 02:49:36 GMT amqp10:connection Transitioning from START to DISCONNECTED due to error
Fri, 20 Nov 2015 02:49:36 GMT amqp10:client connecting to: amqp://localhost
Fri, 20 Nov 2015 02:49:36 GMT amqp10:connection Connecting to localhost:5672 via straight-up sockets
Fri, 20 Nov 2015 02:49:36 GMT amqp10:connection Transitioning from DISCONNECTED to START due to connect
Fri, 20 Nov 2015 02:49:37 GMT amqp10:connection Error from socket: Error: connect ECONNREFUSED
Fri, 20 Nov 2015 02:49:37 GMT amqp10:client connection error:  { [Error: connect ECONNREFUSED]
  code: 'ECONNREFUSED',
  errno: 'ECONNREFUSED',
  syscall: 'connect' }
Fri, 20 Nov 2015 02:49:37 GMT amqp10:client disconnected
Fri, 20 Nov 2015 02:49:37 GMT amqp10:connection Transitioning from START to DISCONNECTED due to error
Fri, 20 Nov 2015 02:49:38 GMT amqp10:client connecting to: amqp://localhost
Fri, 20 Nov 2015 02:49:38 GMT amqp10:connection Connecting to localhost:5672 via straight-up sockets
Fri, 20 Nov 2015 02:49:38 GMT amqp10:connection Transitioning from DISCONNECTED to START due to connect
Fri, 20 Nov 2015 02:49:39 GMT amqp10:connection Error from socket: Error: connect ECONNREFUSED
Fri, 20 Nov 2015 02:49:39 GMT amqp10:client connection error:  { [Error: connect ECONNREFUSED]
  code: 'ECONNREFUSED',
  errno: 'ECONNREFUSED',
  syscall: 'connect' }
Fri, 20 Nov 2015 02:49:39 GMT amqp10:client disconnected
Fri, 20 Nov 2015 02:49:39 GMT amqp10:connection Transitioning from START to DISCONNECTED due to error
Fri, 20 Nov 2015 02:49:40 GMT amqp10:client connecting to: amqp://localhost
Fri, 20 Nov 2015 02:49:40 GMT amqp10:connection Connecting to localhost:5672 via straight-up sockets
Fri, 20 Nov 2015 02:49:40 GMT amqp10:connection Transitioning from DISCONNECTED to START due to connect
Fri, 20 Nov 2015 02:49:41 GMT amqp10:connection Error from socket: Error: connect ECONNREFUSED
Fri, 20 Nov 2015 02:49:41 GMT amqp10:client connection error:  { [Error: connect ECONNREFUSED]
  code: 'ECONNREFUSED',
  errno: 'ECONNREFUSED',
  syscall: 'connect' }
Fri, 20 Nov 2015 02:49:41 GMT amqp10:client disconnected
Fri, 20 Nov 2015 02:49:41 GMT amqp10:connection Transitioning from START to DISCONNECTED due to error
Fri, 20 Nov 2015 02:49:43 GMT amqp10:client connecting to: amqp://localhost
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Connecting to localhost:5672 via straight-up sockets
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Transitioning from DISCONNECTED to START due to connect
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Sending Header 414d515000010000
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Transitioning from START to HDR_SENT due to connected
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Rx: 414d515000010000
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Server AMQP Version: 414d515000010000 vs 414d515000010000
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Transitioning from HDR_SENT to HDR_EXCH due to validVersion
Fri, 20 Nov 2015 02:49:43 GMT amqp10:framing:amqp Encoding performative: {"descriptor":{"buffer":[0,0,0,0,0,0,0,16],"offset":0},"value":{"id":"conn1447987783644","hostname":"localhost","maxFrameSize":{"typeName":"uint","value":4294967295},"channelMax":{"typeName":"ushort","value":65535},"idleTimeout":{"typeName":"uint","value":120000},"outgoingLocales":{"contents":"en-US"},"incomingLocales":{"contents":"en-US"},"offeredCapabilities":null,"desiredCapabilities":null,"properties":{},"encodeOrdering":["id","hostname","maxFrameSize","channelMax","idleTimeout","outgoingLocales","incomingLocales","offeredCapabilities","desiredCapabilities","properties"]}}
Fri, 20 Nov 2015 02:49:43 GMT amqp10:framing Sending frame: OpenFrame: [0,0,0,76,2,0,0,0,0,83,16,192,63,10,161,17,99,111,110,110,49,52,52,55,57,56,55,55,56,51,54,52,52,161,9,108,111,99,97,108,104,111,115,116,112,255,255,255,255,96,255,255,112,0,1,212,192,163,5,101,110,45,85,83,163,5,101,110,45,85,83,64,64,193,1,0]: 0000004c02000000005310c03f0aa111636f6e6e31343437393837373833363434a1096c6f63616c686f737470ffffffff60ffff700001d4c0a305656e2d5553a305656e2d55534040c10100
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Transitioning from HDR_EXCH to OPEN_SENT due to sendOpen
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Rx: 000000a602000000005310c0990aa100a10070ffffffff607fff7000003a984040e01201a30f414e4f4e594d4f55532d52454c415940c16e0aa30770726f64756374a1084163746976654d51a30c746f7069632d707265666978a108746f7069633a2f2fa30c71756575652d707265666978a10871756575653a2f2fa30776657273696f6ea106352e31322e31a308706c6174666f726da10d4a6176612f312e382e305f3630
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Transitioning from OPEN_SENT to OPENED due to openReceived
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Connected with params { maxFrameSize=4294967295, channelMax=32767, idleTimeout=15000 }
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Setting heartbeat check timeout to 1875
Fri, 20 Nov 2015 02:49:43 GMT amqp10:client connected
Fri, 20 Nov 2015 02:49:43 GMT amqp10:session Transitioning from UNMAPPED to BEGIN_SENT due to sendBegin
Fri, 20 Nov 2015 02:49:43 GMT amqp10:framing:amqp Encoding performative: {"descriptor":{"buffer":[0,0,0,0,0,0,0,17],"offset":0},"value":{"remoteChannel":null,"nextOutgoingId":{"typeName":"uint","value":1},"incomingWindow":{"typeName":"uint","value":100},"outgoingWindow":{"typeName":"uint","value":100},"handleMax":{"typeName":"uint","value":4294967295},"offeredCapabilities":null,"desiredCapabilities":null,"properties":{},"encodeOrdering":["remoteChannel","nextOutgoingId","incomingWindow","outgoingWindow","handleMax","offeredCapabilities","desiredCapabilities","properties"]}}
Fri, 20 Nov 2015 02:49:43 GMT amqp10:framing Sending frame: BeginFrame: [0,0,0,31,2,0,0,1,0,83,17,192,18,8,64,82,1,82,100,82,100,112,255,255,255,255,64,64,193,1,0]: 0000001f02000001005311c012084052015264526470ffffffff4040c10100
Fri, 20 Nov 2015 02:49:43 GMT amqp10:connection Rx: 0000001a02000000005311c00d0560000152014343700000ffff0000001802000000005313c00b045201707fffffff520143
Fri, 20 Nov 2015 02:49:43 GMT amqp10:session processing frame(BeginFrame): {"frameType":0,"channel":0,"remoteChannel":1,"nextOutgoingId":1,"incomingWindow":0,"outgoingWindow":0,"handleMax":65535,"properties":{}}
Fri, 20 Nov 2015 02:49:43 GMT amqp10:session Transitioning from BEGIN_SENT to MAPPED due to beginReceived
Fri, 20 Nov 2015 02:49:43 GMT amqp10:session On BEGIN_RCVD, setting params to (1,0,0,65535)
Fri, 20 Nov 2015 02:49:43 GMT amqp10:client mapped
Fri, 20 Nov 2015 02:49:43 GMT amqp10:session processing frame(FlowFrame): {"frameType":0,"channel":0,"nextIncomingId":1,"incomingWindow":2147483647,"nextOutgoingId":1,"outgoingWindow":0,"drain":false,"echo":false,"properties":{}}
Fri, 20 Nov 2015 02:49:43 GMT amqp10:session New Incoming Window (known id): 2147483647 = 1 + 2147483647 - 1
Fri, 20 Nov 2015 02:49:51 GMT amqp10:framing Sending frame: HeartbeatFrame: [0,0,0,8,2,0,0,0]: 0000000802000000

**Comment by mbroadst** [comment]
@kolonse one more question: what broker are you running against? (version as well please)

**Comment by mbroadst** [comment]
@kolonse hmm very strange, I can see right at the end there that it says:
```
Fri, 20 Nov 2015 02:49:43 GMT amqp10:client mapped
```
This indicates that the connect promise should be resolved properly. Could you post the source code for `testLogClient.js`? Maybe there's a typo there we can help sort out for you.

**Comment by kolonse** [comment]
@mbroadst  OK! testLogClient.js is a test file.  MsgSenderClient.js is my real code! 

`MsgSenderClient.js:`
'use strict';
var AMQPClient = require("amqp10").Client;
console.debug = console.trace;
var MsgSenderClient = function(opt){
    opt = opt || {
        reconnect:{
            forever:true,
            strategy: 'fibonacci',
            retries:10
        }
    } ;
    this.opt = opt ;
    this.logger = opt.logger || console ;
    this.uri = opt.uri || "amqp://localhost" ;
    this.name = opt.name || "amq.msg.queue" ;
    this.client = new AMQPClient();
    this.sender = null ;
}

MsgSenderClient.prototype.run = function(callback){
    var self = this ;
    callback = callback || function(){} ;
    self.client.connect(this.uri)
        .then(function(){
            // 创建接收者
            return self.client.createSender(self.name);
        })
        .then(function(sender){
            self.sender = sender ;
            callback(null);
        })
        .catch(function(err){
            // 出现错误那么要通知用户层
            callback(err);
        });
}

MsgSenderClient.prototype.send = function(message,callback){
    callback = callback || function(){} ;
    var self = this ;
    if(self.sender){
        self.sender.send(message)
            .then(function(){
                callback(null);
            })
            .then(function(err){
                callback(err)
            });
    }else{
        callback(new Error("当前发送器未创建成功"));
    }
}

MsgSenderClient.prototype.sendMsg = function(event,data,callback){
    var message = {
        event:event,
        createAt:new Date(),
        data:data
    }
    this.send(message,callback);
}

module.exports = function(opt){
    var ret = new MsgSenderClient(opt);
    //ret.run();
    return ret ;
}

`testLogClient.js:`

var Client = require("./module/MsgSenderClient/MsgSenderClient.js");
var Promise = require("promise");
var moment = require("moment");
var client = Client();
console.log("run");
client.run(function(err){
    console.log("go");
    if(err) {
        console.log(err);
        return;
    }
    var message = {event:"test",data:{
        uid:30078,
        itemID:2001,
        sysID:1,
        type:0,
        num:1000,
        time:moment().format("YYYY-MM-DD HH:mm:ss")
    }};
    setTimeout(function(){
        var arr = [];
        function job(msg){
            return new Promise(function(fulfill,reject){
                client.send(msg,function(err){
                    if(err) reject(err);
                    else fulfill();
                })
            });
        }
        var count = 1 ;
        var Times = 1 ;
        for(var j = 0; j < count; j ++){
            for( var i = 0; i < Times; i ++ ){
                arr.push(job(message));
            }
            var b = new Date().getTime();
            Promise.all(arr).then(function(){
                var time = new Date().getTime() - b ;
                console.log("Cost Time All: ",time / 1000 ," s");
                console.log("Cost Time All: ",Times * 1000 / time, " 个/s");
            }).catch(function(){
                console.log("error");
            })
        }
    },20000);
})

**Comment by noodlefrenzy** [comment]
```
MsgSenderClient.prototype.send = function(message,callback){
  callback = callback || function(){} ;
  var self = this ;
  if(self.sender){
    self.sender.send(message)
      .then(function(){
        callback(null);
      })
      .then(function(err){
        callback(err)
      });
  }else{
    callback(new Error("当前发送器未创建成功"));
  }
}
```

That second `then(function(err)...` should be `catch(function(err)...`, but other than that the code seems reasonable.

So if you set a breakpoint or add a `console.log` before `createSender` here:
```
MsgSenderClient.prototype.run = function(callback){
  var self = this ;
  callback = callback || function(){} ;
  self.client.connect(this.uri)
    .then(function(){
// 创建接收者
      return self.client.createSender(self.name);
``` 

it never gets hit? @mbroadst is right, if you look at the logs your session is getting mapped, and we're even getting a flow frame back from the server, so we should be resolving the promise.

**Comment by mbroadst** [comment]
yeah that `then` should be a `catch`, but even more confusingly in `MsgSenderClient.prototype.sendMsg` you seem to want a "callback" API for `AMQPClient.send`, but then in `testLogClient.js` you wrap the newly made callback API with another promise! (see: `function job(msg)`) Why not just use the original promise offered by the `amqp10` API? I'll try to run this locally and get a cleaned up version, there are a few more tricks you can use if you indeed want a callback api (and perhaps something we can expose if enough people are interested in it)

**Comment by mbroadst** [comment]
@kolonse I've run your code locally against qpidd and even without changing that extra `then` to a `catch` it still works for me.

If you're using a broker we don't currently test against then you might have uncovered a bug (either on our side, or the broker's side), so it would be helpful to provide more information about what broker you're using, and how it's configured.

**Comment by kolonse** [comment]
@mbroadst   I use apache-activemq for broker. config is default!

**Comment by kolonse** [comment]
@noodlefrenzy   Thanks for you ! You are right!  I will modify it!

**Comment by mbroadst** [comment]
@kolonse hm I'm not sure what to say. I just downloaded ActiveMQ version 5.12.1, and ran it from the downloads folder completely unmodified. Then I ran your code, also completely unmodified, and here was the output:

```
mbroadst@retinoid:issue190 (remove-lodash)$ DEBUG=amqp* node testLogClient.js
run
  amqp10:client connecting to: amqp://192.168.1.9 +0ms
  amqp10:connection Connecting to 192.168.1.9:5672 via straight-up sockets +9ms
  amqp10:connection Transitioning from DISCONNECTED to START due to connect +1ms
  amqp10:connection Sending Header 414d515000010000 +3ms
  amqp10:connection Transitioning from START to HDR_SENT due to connected +0ms
  amqp10:connection Rx: 414d515000010000 +184ms
  amqp10:connection Server AMQP Version: 414d515000010000 vs 414d515000010000 +1ms
  amqp10:connection Transitioning from HDR_SENT to HDR_EXCH due to validVersion +0ms
  amqp10:framing:amqp Encoding performative: {"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,16]},"offset":0},"value":{"id":"conn1448120733950","hostname":"192.168.1.9","maxFrameSize":{"typeName":"uint","value":4294967295},"channelMax":{"typeName":"ushort","value":65535},"idleTimeout":{"typeName":"uint","value":120000},"outgoingLocales":{"contents":"en-US"},"incomingLocales":{"contents":"en-US"},"offeredCapabilities":null,"desiredCapabilities":null,"properties":{},"encodeOrdering":["id","hostname","maxFrameSize","channelMax","idleTimeout","outgoingLocales","incomingLocales","offeredCapabilities","desiredCapabilities","properties"]}} +2ms
  amqp10:framing Sending frame: OpenFrame: {"type":"Buffer","data":[0,0,0,78,2,0,0,0,0,83,16,192,65,10,161,17,99,111,110,110,49,52,52,56,49,50,48,55,51,51,57,53,48,161,11,49,57,50,46,49,54,56,46,49,46,57,112,255,255,255,255,96,255,255,112,0,1,212,192,163,5,101,110,45,85,83,163,5,101,110,45,85,83,64,64,193,1,0]}: 0000004e02000000005310c0410aa111636f6e6e31343438313230373333393530a10b3139322e3136382e312e3970ffffffff60ffff700001d4c0a305656e2d5553a305656e2d55534040c10100 +1ms
  amqp10:connection Transitioning from HDR_EXCH to OPEN_SENT due to sendOpen +1ms
  amqp10:connection Rx: 000000a602000000005310c0990aa100a10070ffffffff607fff7000003a984040e01201a30f414e4f4e594d4f55532d52454c415940c16e0aa30770726f64756374a1084163746976654d51a308706c6174666f726da10d4a6176612f312e372e305f3835a30c746f7069632d707265666978a108746f7069633a2f2fa30c71756575652d707265666978a10871756575653a2f2fa30776657273696f6ea106352e31322e31 +21ms
  amqp10:connection Transitioning from OPEN_SENT to OPENED due to openReceived +3ms
  amqp10:connection Connected with params { maxFrameSize=4294967295, channelMax=32767, idleTimeout=15000 } +0ms
  amqp10:connection Setting heartbeat check timeout to 1875 +1ms
  amqp10:client connected +0ms
  amqp10:session Transitioning from UNMAPPED to BEGIN_SENT due to sendBegin +2ms
  amqp10:framing:amqp Encoding performative: {"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,17]},"offset":0},"value":{"remoteChannel":null,"nextOutgoingId":{"typeName":"uint","value":1},"incomingWindow":{"typeName":"uint","value":100},"outgoingWindow":{"typeName":"uint","value":100},"handleMax":{"typeName":"uint","value":4294967295},"offeredCapabilities":null,"desiredCapabilities":null,"properties":{},"encodeOrdering":["remoteChannel","nextOutgoingId","incomingWindow","outgoingWindow","handleMax","offeredCapabilities","desiredCapabilities","properties"]}} +0ms
  amqp10:framing Sending frame: BeginFrame: {"type":"Buffer","data":[0,0,0,31,2,0,0,1,0,83,17,192,18,8,64,82,1,82,100,82,100,112,255,255,255,255,64,64,193,1,0]}: 0000001f02000001005311c012084052015264526470ffffffff4040c10100 +1ms
  amqp10:connection Rx: 0000001a02000000005311c00d0560000152014343700000ffff0000001802000000005313c00b045201707fffffff520143 +8ms
  amqp10:session processing frame(BeginFrame): {"frameType":0,"channel":0,"remoteChannel":1,"nextOutgoingId":1,"incomingWindow":0,"outgoingWindow":0,"handleMax":65535,"properties":{}} +1ms
  amqp10:session Transitioning from BEGIN_SENT to MAPPED due to beginReceived +0ms
  amqp10:session On BEGIN_RCVD, setting params to (1,0,0,65535) +0ms
  amqp10:client mapped +1ms
  amqp10:session processing frame(FlowFrame): {"frameType":0,"channel":0,"nextIncomingId":1,"incomingWindow":2147483647,"nextOutgoingId":1,"outgoingWindow":0,"drain":false,"echo":false,"properties":{}} +2ms
  amqp10:session New Incoming Window (known id): 2147483647 = 1 + 2147483647 - 1 +0ms
  amqp10:link undefined:0: Transitioning from DETACHED to ATTACHING due to sendAttach +3ms
  amqp10:link attach CH=1, Handle=0 +1ms
  amqp10:framing:amqp Encoding performative: {"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,18]},"offset":0},"value":{"name":"amq.msg.queue_4fa3c456-ec7a-4744-ad3e-12c4f0b864fb","handle":{"typeName":"uint","value":0},"role":false,"senderSettleMode":{"typeName":"ubyte","value":2},"receiverSettleMode":{"typeName":"ubyte","value":0},"source":{"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,40]},"offset":0},"address":"localhost","durable":0,"expiryPolicy":"session-end","timeout":0,"dynamic":false},"target":{"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,41]},"offset":0},"address":"amq.msg.queue","durable":0,"expiryPolicy":{"contents":"session-end"},"timeout":0,"dynamic":false},"unsettled":{},"incompleteUnsettled":false,"initialDeliveryCount":{"typeName":"uint","value":1},"maxMessageSize":{"typeName":"ulong","value":0},"offeredCapabilities":null,"desiredCapabilities":null,"properties":{},"encodeOrdering":["name","handle","role","senderSettleMode","receiverSettleMode","source","target","unsettled","incompleteUnsettled","initialDeliveryCount","maxMessageSize","offeredCapabilities","desiredCapabilities","properties"]}} +1ms
  amqp10:framing Sending frame: AttachFrame: {"type":"Buffer","data":[0,0,0,168,2,0,0,1,0,83,18,192,155,14,161,50,97,109,113,46,109,115,103,46,113,117,101,117,101,95,52,102,97,51,99,52,53,54,45,101,99,55,97,45,52,55,52,52,45,97,100,51,101,45,49,50,99,52,102,48,98,56,54,52,102,98,67,66,80,2,80,0,0,83,40,192,38,11,161,9,108,111,99,97,108,104,111,115,116,67,163,11,115,101,115,115,105,111,110,45,101,110,100,67,66,193,1,0,64,193,1,0,64,64,64,0,83,41,192,36,7,161,13,97,109,113,46,109,115,103,46,113,117,101,117,101,67,163,11,115,101,115,115,105,111,110,45,101,110,100,67,66,193,1,0,64,193,1,0,66,82,1,68,64,64,193,1,0]}: 000000a802000001005312c09b0ea132616d712e6d73672e71756575655f34666133633435362d656337612d343734342d616433652d313263346630623836346662434250025000005328c0260ba1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100404040005329c02407a10d616d712e6d73672e717565756543a30b73657373696f6e2d656e644342c1010040c10100425201444040c10100 +34ms
  amqp10:connection Rx: 0000009802000000005312c08b07a132616d712e6d73672e71756575655f34666133633435362d656337612d343734342d616433652d313263346630623836346662434150025000005328c02308a1096c6f63616c686f737443a30b73657373696f6e2d656e644342c1010040c10100005329c02306a10d616d712e6d73672e717565756543a30b73657373696f6e2d656e644342c101000000002002000000005313c013075201707fffffff52014343520170000003e8 +44ms
  amqp10:session processing frame(AttachFrame): {"frameType":0,"channel":0,"name":"amq.msg.queue_4fa3c456-ec7a-4744-ad3e-12c4f0b864fb","handle":0,"role":true,"senderSettleMode":2,"receiverSettleMode":0,"source":{"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,40]},"offset":0},"address":"localhost","durable":0,"expiryPolicy":{"contents":"session-end"},"timeout":0,"dynamic":false,"dynamicNodeProperties":{},"distributionMode":null,"filter":{}},"target":{"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,41]},"offset":0},"address":"amq.msg.queue","durable":0,"expiryPolicy":{"contents":"session-end"},"timeout":0,"dynamic":false,"dynamicNodeProperties":{}},"unsettled":{},"incompleteUnsettled":false,"maxMessageSize":0,"properties":{}} +1ms
  amqp10:link amq.msg.queue_4fa3c456-ec7a-4744-ad3e-12c4f0b864fb:0: Transitioning from ATTACHING to ATTACHED due to attachReceived +0ms
  amqp10:link amq.msg.queue_4fa3c456-ec7a-4744-ad3e-12c4f0b864fb: attached CH=[1=>0], Handle=[0=>0] +0ms
  amqp10:session processing frame(FlowFrame): {"frameType":0,"channel":0,"nextIncomingId":1,"incomingWindow":2147483647,"nextOutgoingId":1,"outgoingWindow":0,"handle":0,"deliveryCount":1,"linkCredit":1000,"drain":false,"echo":false,"properties":{}} +13ms
  amqp10:session New Incoming Window (known id): 2147483647 = 1 + 2147483647 - 1 +1ms
  amqp10:link:sender setting credits (1000,2147483647) +1ms
go
  amqp10:framing Sending frame: HeartbeatFrame: {"type":"Buffer","data":[0,0,0,8,2,0,0,0]}: 0000000802000000 +9s
  amqp10:framing Sending frame: HeartbeatFrame: {"type":"Buffer","data":[0,0,0,8,2,0,0,0]}: 0000000802000000 +8s
  amqp10:link:sender canSend(1000,2147483647) = true +3s
  amqp10:link:sender sending:  +1ms { event: 'test',
  data:
   { uid: 30078,
     itemID: 2001,
     sysID: 1,
     type: 0,
     num: 1000,
     time: Sat Nov 21 2015 10:45:34 GMT-0500 (EST) } }
  amqp10:framing:amqp Encoding performative: {"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,20]},"offset":0},"value":{"handle":{"typeName":"uint","value":0},"deliveryId":{"typeName":"uint","value":1},"deliveryTag":{"type":"Buffer","data":[49]},"messageFormat":{"typeName":"uint","value":0},"settled":false,"more":false,"receiverSettleMode":{"typeName":"ubyte","value":null},"state":null,"resume":false,"aborted":false,"batchable":false,"encodeOrdering":["handle","deliveryId","deliveryTag","messageFormat","settled","more","receiverSettleMode","state","resume","aborted","batchable"]}} +4ms
  amqp10:framing Sending frame: TransferFrame: {"type":"Buffer","data":[0,0,0,121,2,0,0,1,0,83,20,192,15,11,67,82,1,160,1,49,67,66,66,64,64,66,66,66,0,83,119,193,88,4,161,5,101,118,101,110,116,161,4,116,101,115,116,161,4,100,97,116,97,193,66,12,161,3,117,105,100,113,0,0,117,126,161,6,105,116,101,109,73,68,113,0,0,7,209,161,5,115,121,115,73,68,84,1,161,4,116,121,112,101,84,0,161,3,110,117,109,113,0,0,3,232,161,4,116,105,109,101,131,0,0,1,81,42,184,210,82]}: 0000007902000001005314c00f0b435201a001314342424040424242005377c15804a1056576656e74a10474657374a10464617461c1420ca103756964710000757ea1066974656d494471000007d1a10573797349445401a104747970655400a1036e756d71000003e8a10474696d6583000001512ab8d252 +2ms
  amqp10:connection Rx: 0000001802000000005315c00b0541520152014100532445 +14ms
  amqp10:session processing frame(DispositionFrame): {"frameType":0,"channel":0,"role":true,"first":1,"last":1,"settled":true,"state":{"descriptor":{"buffer":{"type":"Buffer","data":[0,0,0,0,0,0,0,36]},"offset":0}},"batchable":false} +1ms
Cost Time All:  0.014  s
```

**Comment by kolonse** [comment]
@mbroadst  Yeah!  If your ActiveMQ is running before the testLogClient run, it's OK! But I want test this case: first, I start `node testLogClient.js` . second, start ActiveMQ!  My question is happen!  Other case is ok!

**Comment by mbroadst** [comment]
@kolonse hey, thanks for the bug report! I think we finally figured this one out, and I hope you don't mind me changing the title of the issue. I've published version `2.1.2` with an included fix for this reconnect issue, please give it a shot

**Comment by kolonse** [comment]
@mbroadst  Very Thanks! It work! Receive my knees!
